### PR TITLE
Initial Draft of Smart Class Parameter TestCases

### PR DESCRIPTION
**Pull request by jyejare**
This is an initial draft of TestCases for feature 'Smart Class Parameters'.

In upstream, Smart Class Parameters appears under following path:
1. Configure -> Classes -> (class_name) -> 'Smart Class Parameter' tab.
2. Configure -> Smart Class Parameters.

### CODE REVIEW

**Comment by oshtaier** [comment]
Don't need `@test: ` word in docstrings anymore per @sthirugn email

**Comment by oshtaier** [comment]
Don't need blank line between:
```
@status: Manual

"""
```

File: tests/foreman/ui/test_classparameters.py [hunk 1]
```diff
@@ -0,0 +1,705 @@
+# -*- encoding: utf-8 -*-
+"""Test class for Smart/Puppet Class Parameter"""
+
+from robottelo.decorators import run_only_on, stubbed, tier3, tier1
+from robottelo.test import UITestCase
+
+
+class ClassParameters(UITestCase):
```
Comment (oshtaier) [inline tests/foreman/ui/test_classparameters.py]: `class SmartClassParametersTestCase` title should be more correct

File: tests/foreman/ui/test_classparameters.py [hunk 2]
```diff
@@ -0,0 +1,705 @@
+# -*- encoding: utf-8 -*-
+"""Test class for Smart/Puppet Class Parameter"""
+
+from robottelo.decorators import run_only_on, stubbed, tier3, tier1
```
Comment (oshtaier) [inline tests/foreman/ui/test_classparameters.py]: `from robottelo.decorators import run_only_on, stubbed, tier1, tier3`

File: tests/foreman/ui/test_classparameters.py [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+from robottelo.test import UITestCase
+
+
+class ClassParameters(UITestCase):
+    """Implements Smart Class Parameter tests in UI"""
+
+    @stubbed()
```
Comment (oshtaier) [inline tests/foreman/ui/test_classparameters.py]: nitpick for all tests:
```
@run_only_on('sat')
@stubbed()
@tier3
```

File: tests/foreman/ui/test_classparameters.py [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+    @run_only_on('sat')
+    @tier3
+    def test_positive_override_checkbox(self):
+        """@test: Override the Default Parameter value.
+
+        @feature: Class Parameters - Override
+
+        @steps:
+        1.  Check an Override checkbox.
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: s/an/the - please update all instances

File: tests/foreman/ui/test_classparameters.py [hunk 5]
```diff
@@ -0,0 +1,705 @@
...
+        2.  Set the new valid Default Value.
+        3.  Submit the changes.
+
+        @assert: Parameter Value overrided with new value.
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: s/overrided/overriden

File: tests/foreman/ui/test_classparameters.py [hunk 6]
```diff
@@ -0,0 +1,705 @@
...
+
+    @stubbed()
+    @run_only_on('sat')
+    @tier3
+    def test_positive_override_checkbox(self):
+        """@test: Override the Default Parameter value.
+
+        @feature: Class Parameters - Override
+
+        @steps:
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: Inorder for the doc to display this correctly, please leave an empty line after this line.

Example:
```py
"""
 @steps:

    1. Check the override checkobx
"""

Please update all instances
```

File: tests/foreman/ui/test_classparameters.py [hunk 7]
```diff
@@ -0,0 +1,705 @@
...
+        1.  Check an Override checkbox.
+        2.  Set the new valid Default Value.
+        3.  Submit the changes.
+
+        @assert: Parameter Value overrided with new value.
+
+        @status: Manual
+
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: This empty line can be removed. Please update all instances.

File: tests/foreman/ui/test_classparameters.py [hunk 8]
```diff
@@ -0,0 +1,705 @@
...
+        """@test: Override the Default Parameter value - override Unchecked.
+
+        @feature: Class Parameters - Override
+
+        @steps:
+        1.  Don't check an Override checkbox.
+        2.  Set the new valid Default Value.
+        3.  Submit the changes.
+
+        @assert: Parameter value not allowed/disbaled to override.
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: s/disbaled/disabled

File: tests/foreman/ui/test_classparameters.py [hunk 9]
```diff
@@ -0,0 +1,705 @@
...
+    @tier3
+    def test_negative_override_checkbox(self):
+        """@test: Override the Default Parameter value - override Unchecked.
+
+        @feature: Class Parameters - Override
+
+        @steps:
+        1.  Don't check an Override checkbox.
+        2.  Set the new valid Default Value.
+        3.  Submit the changes.
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: s/Submit the changes/Attempt to submit the changes.

File: tests/foreman/ui/test_classparameters.py [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+    @tier1
+    def test_positive_update_parameter_type_string(self):
+        """@test: Parameter Update for string type.
+
+        @feature: Class Parameters - Parameter Type
+
+        @steps:
+        1.  Check an Override checkbox.
+        2.  Update the Key Type to String.
+        3.  Give a 'valid' default Value.
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: s/Give/Enter

File: tests/foreman/ui/test_classparameters.py [hunk 11]
```diff
@@ -0,0 +1,705 @@
...
+        """@test: Validation doesn't works on puppet default value.
+
+        @feature: Class Parameters - Optional Validation
+
+        @steps:
+        1.  Check an Override checkbox.
+        2.  Check 'Use Puppet Default' value.
+        3.  Validate this value under section 'Optional Input Validator'.
+
+        @assert: Validation doesn't works with puppet default value.
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: s/works/work

File: tests/foreman/ui/test_classparameters.py [hunk 12]
```diff
@@ -0,0 +1,705 @@
...
+
+        @feature: Class Parameters - Optional Validation
+
+        @steps:
+        1.  Check an Override checkbox.
+        2.  Don't provide any default value, keep blank.
+        3.  Check Required checkbox in 'Optional Input Validator'.
+        4.  Submit the change.
+
+        @assert: Error raised for blank default value by 'Required' checkbox.
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: Are you sure, you expecting an error for a positive test case?

File: tests/foreman/ui/test_classparameters.py [hunk 13]
```diff
@@ -0,0 +1,705 @@
...
+
+        @feature: Class Parameters - Optional Validation
+
+        @steps:
+        1.  Check an Override checkbox.
+        2.  Provide some default value.
+        3.  Check Required checkbox in 'Optional Input Validator'.
+        4.  Submit the change.
+
+        @assert: Error not raised default value by 'Required' checkbox.
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: Should error be thrown here?

File: tests/foreman/ui/test_classparameters.py [hunk 14]
```diff
@@ -0,0 +1,705 @@
...
+
+        @status: Manual
+
+        """
+
+    @stubbed()
+    @run_only_on('sat')
+    @tier3
+    def test_positive_validate_matcher_value_required_checkbox(self):
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: I think the positive and negative test cases got interchanged here.

File: tests/foreman/ui/test_classparameters.py [hunk 15]
```diff
@@ -0,0 +1,705 @@
...
+        @assert: Error not raised for matcher value by 'Required' checkbox.
+
+        @status: Manual
+
+        """
+
+    @stubbed()
+    @run_only_on('sat')
+    @tier3
+    def test_positive_validate_default_value_with_regex(self):
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: I think the positive and negative test cases got interchanged here.

**Comment by sthirugn** [comment]
Is it necessary to create separate positive test cases for each parameter types :
* string
* boolean
* integer
* real
* array
* hash
* yaml
* json

Can't we just create one positive and one negative test case and cover all parameter types?

**Comment by sthirugn** [comment]
@jyejare also can you please ping me in the morning, I have few things to talk to you.

File: tests/foreman/ui/test_classparameters.py (hunk 1 above)
Comment (jyejare) [inline tests/foreman/ui/test_classparameters.py]: Renamed.

File: tests/foreman/ui/test_classparameters.py (hunk 3 above)
Comment (jyejare) [inline tests/foreman/ui/test_classparameters.py]: Done.

File: tests/foreman/ui/test_classparameters.py (hunk 6 above)
Comment (jyejare) [inline tests/foreman/ui/test_classparameters.py]: Done.

File: tests/foreman/ui/test_classparameters.py (hunk 4 above)
Comment (jyejare) [inline tests/foreman/ui/test_classparameters.py]: Done.

**Comment by jyejare** [comment]
@oshtaier All your comments are addressed. Please take a second look.
@sthirugn : All your comments are addressed. Also I added one positive and one negative test case for all the parameter types.

Thanks

File: tests/foreman/ui/test_classparameters.py (hunk 12 above)
Comment (jyejare) [inline tests/foreman/ui/test_classparameters.py]: This is been addressed by changing the Test Case name to negative and below one to negative.

File: tests/foreman/ui/test_classparameters.py [hunk 16]
```diff
@@ -0,0 +1,469 @@
+# -*- encoding: utf-8 -*-
+"""Test class for Smart/Puppet Class Parameter"""
+
+from robottelo.decorators import run_only_on, stubbed, tier1, tier3
+from robottelo.test import UITestCase
+
+
+class SmartClassParameters(UITestCase):
```
Comment (oshtaier) [inline tests/foreman/ui/test_classparameters.py]: No-no.
Please use `SmartClassParametersTestCase` name
That is a naming convention as per https://github.com/SatelliteQE/robottelo/issues/3102

File: tests/foreman/ui/test_classparameters.py [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+    """Implements Smart Class Parameter tests in UI"""
+
+    @stubbed()
```
Comment (oshtaier) [inline tests/foreman/ui/test_classparameters.py]: Probably, forgot sorting there?

**Comment by oshtaier** [comment]
ACK pending comments

File: tests/foreman/ui/test_classparameters.py (hunk 16 above)
Comment (jyejare) [inline tests/foreman/ui/test_classparameters.py]: I forgot to concatenate TestCase word. :)

**Comment by oshtaier** [comment]
ACK

File: tests/foreman/ui/test_classparameters.py [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_update_parameter_type(self):
+        """Positive Parameter Update for parameter types - Valid Value.
+
+        Types: string, boolean, integer, real, array, hash, yaml, json
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: Testimony currently doesn't allow : in test case names. Please 

s/'Types:'/'Types - '

File: tests/foreman/ui/test_classparameters.py [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_negative_update_parameter_type(self):
+        """Negative Parameter Update for parameter types - Invalid Value.
+
+        Types: string, boolean, integer, real, array, hash, yaml, json
```
Comment (sthirugn) [inline tests/foreman/ui/test_classparameters.py]: Testimony currently doesn't allow : in test case names. Please

s/'Types:'/'Types - '

**Comment by sthirugn** [comment]
Pending testimony failure in travis

**Comment by sthirugn** [comment]
ACK

**Pull request merged**